### ISSUE TITLE
Nextjs with jest example will not work if there is an esm dependency in node_modules

### ISSUE THREAD

**Issue by wmira** [vercel/next.js#40183]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Fri Apr 2 22:23:49 UTC 2021
    Binaries:
      Node: 16.16.0
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 12.2.5
      eslint-config-next: 12.2.5
      react: 18.2.0
      react-dom: 18.2.0

### Which example does this report relate to?

https://nextjs.org/docs/testing

### Describe the Bug

If you have a project with a dependency that is published as an es modules in node_modules, running jest will not work..even though nextjs works with it e.g. running the project normally with npm run dev.

Test does not run even if ignoring the package. Please check this repo that reproduce the problem:

https://github.com/wmira/banana/

if you run npm run dev here..works fine.. npm run test here does not and this follow the jest example.

### Expected Behavior

running test via jest works as per the docs

### To Reproduce

repo is below:
https://github.com/wmira/banana/

1. npm run dev -- nextjs works fine with an es module dependency under node_modules
2. npm run test -- throws error

**Comment by bruceharrison1984** [comment]
I ran into this issue once I installed preact in to my project.

**Comment by edwinfernaldy** [comment]
im getting this error too in my project.

**Comment by dariolongo** [comment]
Same issue for me.
I've fixed with:
```js
module.exports = async () => ({
  ...(await createJestConfig(customJestConfig)()),
  transformIgnorePatterns: ["node_modules/(?!(package1|package2)/)", "^.+\\.module\\.(css|sass|scss)$"],
});
```
In my case packages are: ```swiper|ssr-window|dom7```
Any suggestion?

**Comment by nora** [comment]
I had the same.

The cause of this issue is that I cannot include any module under node_modules in Jest's trasform target.

Specifically,

- `/node_modules/` is defined in `transformIgnorePatterns` by `next/jest`.
- customJestConfig cannot override it and is limited to adding it

This is the relevant part of next/jest.
https://github.com/vercel/next.js/blob/v12.3.0/packages/next/build/jest/jest.ts#L136

The solution offered by @dariolongo is adequate.
However, instead of completely overriding it, a diversion of the standard ignorePatterns might be safer.

```js
const esModules = ['package1', 'package2']
const customConfig = {
  // Dependency packages must also be included in this list
  transformIgnorePatterns: [`/node_modules/(?!(${esModules.join('|')})/)`],
  // And other custom config...
}

module.exports = async () => {
  const jestConfig = await _createJestConfig(customConfig)()
  return {
    ...jestConfig,
    transformIgnorePatterns: jestConfig.transformIgnorePatterns.filter(
      (ptn) => ptn !== '/node_modules/'
    ), // ['^.+\\.module\\.(css|sass|scss)$', '/node_modules/(?!(package1|package2)/']
  }
}
```

I would like to suggest that the documentation be supplemented with "config cannot be completely overwritten".

**Comment by ThisIsMissEm** [comment]
I've been trying to look into the next.js configuration, and from what I can see, `node_modules` isn't excluded in the standard next.js config when using esm and swc (both defaults as of Next 12), so I'm suspecting that this might just be a left over from a pre-ESM time?

**Comment by BoilingSoup** [comment]
Solution from @nora worked perfectly. I believe this is a bug, but at the very least the docs should be updated to show how `transformIgnorePatterns` can be configured correctly. What's currently implied in the docs is incorrect.

**Comment by oosawy** [comment]
> The solution offered by @dariolongo is adequate. However, instead of completely overriding it, a diversion of the standard ignorePatterns might be safer.
> 
> ```js
> const esModules = ['package1', 'package2']
> const customConfig = {
>   // Dependency packages must also be included in this list
>   transformIgnorePatterns: [`/node_modules/(?!(${esModules.join('|')})/)`],
>   // And other custom config...
> }
> 
> module.exports = async () => {
>   const jestConfig = await _createJestConfig(customConfig)()
>   return {
>     ...jestConfig,
>     transformIgnorePatterns: jestConfig.transformIgnorePatterns.filter(
>       (ptn) => ptn !== '/node_modules/'
>     ), // ['^.+\\.module\\.(css|sass|scss)$', '/node_modules/(?!(package1|package2)/']
>   }
> }
> ```
> 
> I would like to suggest that the documentation be supplemented with "config cannot be completely overwritten".

When using Next.js 13.1 [transpilePackages](https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable), this solution seems not working by [another regex next/jest added](https://github.com/vercel/next.js/blob/v13.1.1/packages/next/build/jest/jest.ts#L158-L163).

But now just put all `esModules` into transpilePackages in next.config.js works perfectly for me!

```js
module.exports = {
  transpilePackages: ['package1', 'package2'],
}
```

**Comment by sshah98** [comment]
fyi i fixed this issue by removing the default paths in tsconfig     "paths": {
      "*": ["./*"]
    }

**Comment by tachun** [comment]
> > The solution offered by @dariolongo is adequate. However, instead of completely overriding it, a diversion of the standard ignorePatterns might be safer.
> > ```js
> > const esModules = ['package1', 'package2']
> > const customConfig = {
> >   // Dependency packages must also be included in this list
> >   transformIgnorePatterns: [`/node_modules/(?!(${esModules.join('|')})/)`],
> >   // And other custom config...
> > }
> > 
> > module.exports = async () => {
> >   const jestConfig = await _createJestConfig(customConfig)()
> >   return {
> >     ...jestConfig,
> >     transformIgnorePatterns: jestConfig.transformIgnorePatterns.filter(
> >       (ptn) => ptn !== '/node_modules/'
> >     ), // ['^.+\\.module\\.(css|sass|scss)$', '/node_modules/(?!(package1|package2)/']
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I would like to suggest that the documentation be supplemented with "config cannot be completely overwritten".
> 
> When using Next.js 13.1 [transpilePackages](https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable), this solution seems not working by [another regex next/jest added](https://github.com/vercel/next.js/blob/v13.1.1/packages/next/build/jest/jest.ts#L158-L163).
> 
> But now just put all `esModules` into transpilePackages in next.config.js works perfectly for me!
> 
> ```js
> module.exports = {
>   transpilePackages: ['package1', 'package2'],
> }
> ```

As mentioned by [@oosawy](https://github.com/oosawy), with the use of Next.js 13.1's [transpilePackages](https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable), the regex has changed and will generate these `transformIgnorePatterns`:
```
module.exports = {
  transpilePackages: ['package1', 'package2'],
}
```
```
[
  '/node_modules/(?!.pnpm)(?!(package1|package2)/)',
  '/node_modules/.pnpm/(?!(package1|package2)@)',
  '^.+\\.module\\.(css|sass|scss)$'
]
```
In this case, we don't need to add `transformIgnorePatterns` in the custom configuration, but, I need to remove the first element from the generated transformIgnorePatterns array to make it work.

```
module.exports = async () => {
  const jestConfig = await _createJestConfig(customConfig)()
  return {
    ...jestConfig,
    transformIgnorePatterns: jestConfig.transformIgnorePatterns.filter(
      (ptn) => ptn !== '/node_modules/(?!.pnpm)(?!(package1|package2)/)'
    ),
  }
}
```

**Comment by drewloomer** [comment]
This workaround seems to be broken with the release of 13.4. Has anyone else had luck upgrading to that release?

**Comment by jonioni** [comment]
Also encountering this issue. Wondering if next jest config could inspect `type: "module"` of the dependencies and auto add them into `transformIgorePatterns`?

**Comment by LarsEjaas** [comment]
I am experiencing the same issue with an ESM node_module causing Jest errors in v.13.4.1 and ended up downgrading to v.13.2.3. Tried different combinations of `transpilePackages` in `next.config.mjs` and `transformIgnorePatterns` in `jest.config.mjs` but finally gave up and downgraded the version.

**Comment by brett-minca** [comment]
I just ran into this bug for the `query-string` module.  I found that adding `query-string` and all of it's dependencies to `transpilePackages` worked.

```next.config.ts
```

**Comment by AlfredMadere** [comment]
Has this been resolved? I am still unable to use Jest with nextjs in 13.4 due to esm issues in node_modules.

**Comment by paulhchoi** [comment]
Dittoing @AlfredMadere, I'm also still unable to use Jest with Next.JS in 13.5.4 due to esm issues.

**Comment by leandroruel** [comment]
i tried everything in this issue and i`m stuck forever here

**Comment by twobit** [comment]
@leandroruel try converting your ESM dependencies for Jest:

```
npx esbuild [YOUR_DEP] --bundle --platform=node --outfile=vendor/[YOUR_DEP].js
```

Then update your `jest.config.js`:
```
moduleNameMapper: {
    '^[YOUR_DEP]$': '<rootDir>/vendor/[YOUR_DEP]',
    ...
}
```

**Comment by atstoyanov** [comment]
I think I'll stick with `babel` until this one is resolved - https://nextjs.org/docs/pages/building-your-application/optimizing/testing#setting-up-jest-with-babel

**Comment by ankita-khandelwal** [comment]
> > The solution offered by @dariolongo is adequate. However, instead of completely overriding it, a diversion of the standard ignorePatterns might be safer.
> > ```js
> > const esModules = ['package1', 'package2']
> > const customConfig = {
> >   // Dependency packages must also be included in this list
> >   transformIgnorePatterns: [`/node_modules/(?!(${esModules.join('|')})/)`],
> >   // And other custom config...
> > }
> > 
> > module.exports = async () => {
> >   const jestConfig = await _createJestConfig(customConfig)()
> >   return {
> >     ...jestConfig,
> >     transformIgnorePatterns: jestConfig.transformIgnorePatterns.filter(
> >       (ptn) => ptn !== '/node_modules/'
> >     ), // ['^.+\\.module\\.(css|sass|scss)$', '/node_modules/(?!(package1|package2)/']
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I would like to suggest that the documentation be supplemented with "config cannot be completely overwritten".
> 
> When using Next.js 13.1 [transpilePackages](https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable), this solution seems not working by [another regex next/jest added](https://github.com/vercel/next.js/blob/v13.1.1/packages/next/build/jest/jest.ts#L158-L163).
> 
> But now just put all `esModules` into transpilePackages in next.config.js works perfectly for me!
> 
> ```js
> module.exports = {
>   transpilePackages: ['package1', 'package2'],
> }
> ```

Above suggestion worked for me with Nextjs 14.0.3.

**Comment by esantosrvolvo** [comment]
I am on NextJS 14.0.3 and the `transpilePackages` solution did not work for me for package `next-image-export-optimizer`.
I got it working by[ switching back to babel](https://nextjs.org/docs/pages/building-your-application/optimizing/testing#setting-up-jest-with-babel) as indicated by @atstoyanov.

**Comment by atstoyanov** [comment]
It work in 14.0.3 using `transpilePackages` and without the need of overriding the `transformIgnorePatterns`.  Here is my configuration.

```javascript
// next.config.js
{
  transpilePackages: [
    'axios',
    'react-dnd-html5-backend',
    'supports-webp',
    'ramda',
    'react-firebase-hooks/auth',
  ]
}

// jest.config.ts
import type { Config } from 'jest';
import nextJest from 'next/jest.js';

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
});

// Add any custom config to be passed to Jest
const config: Config = {
  coverageProvider: 'v8',
  testEnvironment: 'jsdom',
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
  moduleNameMapper: {
    'react-markdown': '<rootDir>/node_modules/react-markdown/react-markdown.min.js',
    '^@/(.*)$': '<rootDir>/$1',
  },
  testPathIgnorePatterns: ['/tests/'],
  resetMocks: false,
  clearMocks: true,
};

export default createJestConfig(config);
```

and here is the produced config:
```javascript
{
  coverageProvider: 'v8',
  testEnvironment: 'jsdom',
  setupFilesAfterEnv: [ '<rootDir>/jest.setup.js' ],
  moduleNameMapper: {
    '^.+\\.module\\.(css|sass|scss)$': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/object-proxy.js',
    '^.+\\.(css|sass|scss)$': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/styleMock.js',
    '^.+\\.(png|jpg|jpeg|gif|webp|avif|ico|bmp)$': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/fileMock.js',
    '^.+\\.(svg)$': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/fileMock.js',
    '@next/font/(.*)': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/nextFontMock.js',
    'next/font/(.*)': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/nextFontMock.js',
    'server-only': '/Users/user/src/demo-app/node_modules/next/dist/build/jest/__mocks__/empty.js',
    'react-markdown': '<rootDir>/node_modules/react-markdown/react-markdown.min.js',
    '^@/(.*)$': '<rootDir>/$1'
  },
  testPathIgnorePatterns: [ '/node_modules/', '/.next/', '/tests/' ],
  resetMocks: false,
  clearMocks: true,
  transform: {
    '^.+\\.(js|jsx|ts|tsx|mjs)$': [
      '/Users/user/src/demo-app/node_modules/next/dist/build/swc/jest-transformer.js',
      [Object]
    ]
  },
  transformIgnorePatterns: [
    '/node_modules/(?!.pnpm)(?!(axios|react-dnd-html5-backend|supports-webp|ramda|react-firebase-hooks/auth)/)',
    '/node_modules/.pnpm/(?!(axios|react-dnd-html5-backend|supports-webp|ramda|react-firebase-hooks\\+auth)@)',
    '^.+\\.module\\.(css|sass|scss)$'
  ],
  watchPathIgnorePatterns: [ '/.next/' ]
}
```

**Comment by RobsonAraujo** [comment]
I'm getting this issue with Next 14.1.0, even when using [create-next-app](https://nextjs.org/docs/app/api-reference/create-next-app). It looks like ESLINT and JEST/Cypress are not working together. 

It is possible to quickly reproduce it following the steps below: 

Run `npx create-next-app@latest` (Choose the option to include Eslint on the config) 

Then, follow the step to add config Jest
`https://nextjs.org/docs/app/building-your-application/testing/jest`

Then, try to run the `yarn test` command. The following bug will be shown
![image](https://github.com/vercel/next.js/assets/17827897/a1a09570-a6cc-4cc3-94e1-676cf78d149b)

I have tried to use `transpilePackages ` as mentioned above but without success. Any idea?

**Comment by leandroruel** [comment]
if it helps someone here's my config file:

```js
const nextJest = require('next/jest')
const swiperMocks = require('./src/__mocks__/misc/swiper.ts')

const createJestConfig = nextJest({
  // Path to Next.js app to load next.config.js
  dir: './',
})

/** @type {import('@jest/types').Config.InitialOptions} */
const customJestConfig = {
  testEnvironment: 'jest-environment-jsdom',
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
  maxWorkers: 4,
  /**
   * Custom config goes here, I am not adding it to keep this example simple.
   * See next/jest examples for more information.
   */
  moduleNameMapper: {
    '^@/components/(.*)$': '<rootDir>/components/$1',
    '^components/(.*)$': '<rootDir>/src/components/$1',
    '^helpers/(.*)$': '<rootDir>/src/helpers/$1',
    '^generated/(.*)$': '<rootDir>/src/generated/$1',
    '^mixins/(.*)$': '<rootDir>/src/mixins/$1',
    '^hooks/(.*)$': '<rootDir>/src/hooks/$1',
    '^mocks/(.*)$': '<rootDir>/src/__mocks__/$1',
    ...swiperMocks.moduleNameMapper,
  },
  transform: {
    ...swiperMocks.transform,
  },
  modulePaths: ['<rootDir>/src/'],
  collectCoverage: true,
  collectCoverageFrom: [
    '<rootDir>/src/components/**/*.tsx', // Cobertura nos arquivos ts e tsx dentro de /src/components
    '!<rootDir>/src/pages/**', // Exclui arquivos de índice
    '!<rootDir>/src/components/**/stories.tsx', // Exclui arquivos de índice
    '!<rootDir>/src/components/**/index.ts',
  ],
  coveragePathIgnorePatterns: ['.*__snapshots__/.*'],
}

const esModules = [
  'ccount',
  'react-markdown',
  'vfile',
  'unist-.+',
  'unified',
  'bail',
  'is-plain-obj',
  'trough',
  'remark',
  'remark-.+',
  'mdast-util-.+',
  'micromark',
  'parse-entities',
  'character-entities',
  'property-information',
  'comma-separated-tokens',
  'hast-util-whitespace',
  'hast-util-to-html',
  'hast-util-sanitize',
  'html-void-elements',
  'space-separated-tokens',
  'decode-named-character-reference',
  'zwitch',
  'longest-streak',
  'stringify-entities',
  'trim-lines',
  'swiper',
  'swiper/react',
  'ssr-window',
  'dom7',
].join('|')

module.exports = async () => ({
  /**
   * Using ...(await createJestConfig(customJestConfig)()) to override transformIgnorePatterns
   * provided byt next/jest.
   *
   * @link https://github.com/vercel/next.js/issues/36077#issuecomment-1096635363
   */
  ...(await createJestConfig(customJestConfig)()),
  /**
   * Swiper uses ECMAScript Modules (ESM) and Jest provides some experimental support for it
   * but "node_modules" are not transpiled by next/jest yet.
   *
   * @link https://github.com/vercel/next.js/issues/36077#issuecomment-1096698456
   * @link https://jestjs.io/docs/ecmascript-modules
   */
  transformIgnorePatterns: [`<rootDir>/node_modules/(?!${esModules})/`],
})
```

**Comment by AtanasChachev** [comment]
> I'm getting this issue with Next 14.1.0, even when using [create-next-app](https://nextjs.org/docs/app/api-reference/create-next-app). It looks like ESLINT and JEST/Cypress are not working together.
> 
> It is possible to quickly reproduce it following the steps below:
> 
> Run `npx create-next-app@latest` (Choose the option to include Eslint on the config)
> 
> Then, follow the step to add config Jest `https://nextjs.org/docs/app/building-your-application/testing/jest`
> 
> Then, try to run the `yarn test` command. The following bug will be shown ![image](https://private-user-images.githubusercontent.com/17827897/307535645-a1a09570-a6cc-4cc3-94e1-676cf78d149b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyMzk1NjgsIm5iZiI6MTcxNTIzOTI2OCwicGF0aCI6Ii8xNzgyNzg5Ny8zMDc1MzU2NDUtYTFhMDk1NzAtYTZjYy00Y2MzLTk0ZTEtNjc2Y2Y3OGQxNDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDA3MjEwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyYTdmODA0MDlmOTVlMWU1NmVjZTUxZGVmMDQyNWZjNDc4ODlmMzY3NTg0ZGE0Nzc5NjZiN2M5YThiY2M4ZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zaMClvZgboH9lcaNSvNfua5JGtWme_mfTQsSS_NZoRc)
> 
> I have tried to use `transpilePackages ` as mentioned above but without success. Any idea?

I was able to fix that by adding 

```
"resolutions": {
    "string-width": "4.2.3",
    "wrap-ansi": "7.0.0"
  }
```
in my `package.json`. 
Not quite sure if it's the best solution but it fixed the problem :)

Make sure you do `rm -rf node_modules` and then re-run `yarn`.

**Comment by RobsonAraujo** [comment]
@AtanasChachev  That works; thanks for sharing this alternative solution 💯

**Comment by zotijs** [comment]
Does anyone have a config that works for esm only modules and pnpm? I am struggling to make `react-markdown` work with jest

**Comment by leandroruel** [comment]
@rafal-r Yeah, i stopped using jest and I'm using vitest from now on. less headaches.

**Comment by max-ilnytskyi** [comment]
I spent half a day trying to fix the ERR_REQUIRE_ESM error. In the end, everything worked for me after adding ts-jest and ts-node to devDependencies

`yarn add --dev ts-node ts-jest`

My `jest.config.ts`
```
import type { Config } from 'jest';
import nextJest from 'next/jest.js';

const createJestConfig = nextJest({
  dir: './',
});

const config: Config = {
  coverageProvider: 'v8',
  testEnvironment: 'jsdom',
};

export default createJestConfig(config);
```

**Comment by dmurvihill** [comment]
On `next 14.2.22`, `jest 29.7.0`, `ts-jest 29.2.5`, the problem seems to be this glob in the default next-jest config, in `dist/build/jest/jest.js`:

```javascript
                transform: {
                    // Use SWC to compile tests
                    "^.+\\.(js|jsx|ts|tsx|mjs)$": [
                        require.resolve("../swc/jest-transformer"),
                        jestTransformerConfig
                    ],
                    // Allow for appending/overriding the default transforms
                    ...resolvedJestConfig.transform || {}
                },
```

Removing the `|mjs` worked for all of my unit tests; I used this pattern to rewrite it in my `jest.config.ts`:

```typescript
import nextJest from "next/jest.js";

const createJestConfig = nextJest({
  dir: "./",
});

const config = {
  // ...
}

function removeEsmFromTransforms(config: JestConfig.Config.InitialOptions): JestConfig.Config.InitialOptions {
  const transform = config.transform || {};
  const key = "^.+\\.(js|jsx|ts|tsx|mjs)$";
  if (transform.hasOwnProperty(key)){
    const v = transform[key] || '';
    delete transform[key];
    transform[key.replace(/\|mjs/, '')] = v;
  }
  return {
    ...config,
    transform,
  };
}

export default () => createJestConfig(config)().then(removeEsmFromTransforms);
```

A bit awkward, but it's the only thing that worked for me.

**Comment by murat-mehmet** [comment]
Generating list of dependent modules was a bit pain in my case so this helped me:

_esModules.js_
```js
const fs = require('fs');
const path = require('path');

// LIST OF ES MODULES
const esModuleRoots = ['rehype-stringify', 'remark-gfm', 'remark-rehype'];

const nodeModulesDir = path.join(__dirname, 'node_modules');

function readPackageJson(dir) {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    return JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  }
  return null;
}

function getDependencies(dir, tree, altDir) {
  let pkg = readPackageJson(altDir);
  if (!pkg) {
    pkg = readPackageJson(dir);
  }
  if (!pkg) return;

  tree[pkg.name] = {
    version: pkg.version,
    esModule: pkg.type === 'module',
    dependencies: {},
  };

  if (!pkg.dependencies) return;

  const deps = Object.keys(pkg.dependencies);
  deps.forEach(depName => {
    const depDir = path.join(nodeModulesDir, depName);
    const altDir = path.join(nodeModulesDir, pkg.name, 'node_modules', depName);
    getDependencies(depDir, tree[pkg.name].dependencies, altDir);
  });
}

function findESModules(tree, name) {
  const modules = [];

  function traverse(node, name) {
    if (node.esModule) {
      modules.push(name);
    }
    for (const dep in node.dependencies) {
      traverse(node.dependencies[dep], dep);
    }
  }

  traverse(tree, name);
  return modules;
}

function main(moduleName) {
  const moduleDir = path.join(nodeModulesDir, moduleName);
  const tree = {};
  getDependencies(moduleDir, tree, moduleDir);
  // console.log('Dependency Tree:', JSON.stringify(tree, null, 2));

  return findESModules(tree[moduleName], moduleName);
}

function onlyUnique(value, index, array) {
  return array.indexOf(value) === index;
}

const allModules = esModuleRoots.map(main).flat().filter(onlyUnique);
// console.log('ES Modules found:', allModules);

module.exports = allModules;
```

_jest.config.js_
```js
```

**Comment by FedorT22** [comment]
Also, `nest/jest` doesn't work for `next-intl`, while using:
```json
"dependencies": {
    "next": "15.3.4",
    "next-intl": "^4.3.1",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
},

"devDependencies": {
    "jest": "^30.0.4",
    "jest-environment-jsdom": "^30.0.4",
    "ts-node": "^10.9.2",
}
```
___

For me it works only if I remove `next/jest` configuration completely and providing the following jest setup:
```tsx
const config: Config = {
  rootDir: '.',
  testEnvironment: 'jest-environment-jsdom',
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],

  // Implementing transformations manually
  transform: {
    '^.+\\.(t|j)sx?$': [
      '@swc/jest',
      {
        jsc: {
          parser: {
            syntax: 'typescript',
            tsx: true,
          },
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
        module: {
          type: 'commonjs',  // Jest runs tests in a CommonJS environment, but next-intl is ESM
        },
      },
    ],
  },

  // Telling Jest to transform these ESM modules
  transformIgnorePatterns: ['/node_modules/(?!next-intl|use-intl)'],
}
```

Solution description:
- At first, I tried to provide `transformIgnorePatterns`. It didn't work.
- Then I added `testPathIgnorePatterns`, hoping it will override next/jest behavior. It didn't work.
- Then, I added `transform` option, still didn't work. It didn't work.

Recap:
I don't use `next/jest` at all by setting up `jest` from scratch and then partially implementing `next/jest` behavior described in official `next.js` documentation https://nextjs.org/docs/app/guides/testing/jest#:~:text=Under%20the%20hood,for%20you%2C%20including%3A.

**Comment by FedorT22** [comment]
After looking at the next.js codebase I found out the reason and the easy fix:
```ts
// /next.config.ts

const nextConfig: NextConfig = {
  transpilePackages: ['next-intl', 'use-intl'],
  ...
}
```

Unfortunately, [documentation](https://nextjs.org/docs/app/api-reference/config/next-config-js/transpilePackages) doesn't say to much about it, so I'll elaborate it a bit in case of our issue:

- `transformIgnorePatterns`: The code constructs this configuration to explicitly ignore transformations for modules inside `node_modules`, unless they are specifically listed in the `transpilePackages` option in the `next.config.ts`.

`next/jest` tool assumes that all packages in `node_modules` are already in a format that `jest` can understand (`CommonJS`). When it encountered `next-intl`, which uses modern `ESM` export syntax, it followed its rules, did not transform the file, and passed the incompatible code directly to`jest`, causing the crash.

In summary, I think,  `next/jest` is intentionally designed to bypass `node_modules` for performance reasons. This optimization becomes a point of failure when a dependency, like `next-intl`, does not ship code in the older `CommonJS` format, that the `jest` test runner expects by default.

See the next/jest source code here:
https://github.com/vercel/next.js/blob/d7fa64ea3babbcdebabeaae48bbc3da3bb347828/packages/next/src/build/jest/jest.ts#L171C9-L171C32
